src/mono/mono/mini/CMakeLists.txt

```diff
@@ -388,6 +388,30 @@ if(NOT DISABLE_SHARED_LIBS)
     add_library(monosgen-shared-dac SHARED "mini-windows-dlldac.c")
     set_target_properties(monosgen-shared-dac PROPERTIES OUTPUT_NAME ${MONO_SHARED_LIB_NAME}-dac)
   endif()
+
+
+  if(TARGET_DARWIN)
+    add_library(mono-framework SHARED $<TARGET_OBJECTS:monosgen-objects>)
+    target_compile_definitions(mono-framework PRIVATE -DMONO_DLL_EXPORT)
+    target_sources(mono-framework PRIVATE $<TARGET_OBJECTS:eglib_objects>)
+    target_link_libraries(mono-framework PRIVATE ${OS_LIBS} ${ICONV_LIB} ${LLVM_LIBS} ${ICU_LIBS})
+    if(ICU_LDFLAGS)
+      set_property(TARGET mono-framework APPEND_STRING PROPERTY LINK_FLAGS " ${ICU_LDFLAGS}")
+    endif()
+    if(STATIC_ICU)
+      set_property(TARGET mono-framework APPEND_STRING PROPERTY LINKER_LANGUAGE CXX)
+    endif ()
+    set_property(TARGET mono-framework APPEND_STRING PROPERTY LINK_FLAGS " -Wl,-compatibility_version -Wl,2.0 -Wl,-current_version -Wl,2.0")
+    target_sources(mono-framework PRIVATE "${mono-components-objects}")
```

Comment:
So this is going to be a mono framework for ios/tvos/catalyst that always statically links in the full non-stub versions of all the components.  Is that what we want?  We always want to support hot reload and diagnostics, @rolfbjarne ?

Or should each component be its own dynamic .framework, too?  (in which case... it would be upto the xamarin-macios embedding code would probably need to hook mono_dl_open and locate the components for us)

Reply:
>  statically links

I have not read the PR yet... but the goal of a `.framework` is that it's all dynamic (not static) linking

Reply:
yea mono is built as a dynamic thing, but it will include all the components statically as part of it.  I left some comments on the Paper.

Reply:
Isn't there a hard limit to the number of .frameworks an app can have?

Reply:
> We always want to support hot reload and diagnostics,

# Debug

* We do **not** care about the app bundle size [1]
  * IOW all features (including HR and diagnostics) can be available in the binary (and be enabled/disabled by the app if required)
* We care about build and deploy times
  * dynamic linking (used for frameworks) makes the build faster
  * the framework (binary) will rarely change (deployed to sim/device once)

[1] unless it starts affecting build/deploy times :-)

# Release

* We care about the app bundle size
   * by default, we won't be using the `.framework`
     * using the static libraries `.a` allows the native linker / strip to remove a lot of code. Using the framework would create much larger apps.
   * a release `.framework` is needed for app extensions
     * why ? it's smaller to share a single, larger `.framework` than to have a (even partial) runtime in 2 (or more) executables - main app and one/several extension(s)
* We do **not** care (within limits) about build times. Release builds being less common than debug ones
* We do **not** care about deploy times. Since release build are smaller then generally deploy faster anyway.

I do not see "hot reload" needed for release builds. Not entirely sure about "diagnostics" ...

Modularity is more important for the static libraries. The majority of apps ships without extensions (but that could change).

> Or should each component be its own dynamic .framework, too?

There's a limit (at least a suggested one, never seen a rejection) to the number of user frameworks that an app should have. However a framework can have several dynamic libraries inside it... so we could control what's in or out.

Not sure if that's really needed...

Reply:
> Isn't there a hard limit to the number of .frameworks an app can have?

@steveisok yes, but not sure how hard it is (or could become).

@rolfbjarne you mentioned that (limit) earlier but I can't find a link, got anything more ?

Reply:
> however a framework can have several dynamic libraries inside it... so we could control what's in or out.

Ok, in that case I think we should have a Mono-framework.framework with `libmonosgen-2.0.dylib` in it and `libmono-component-hot_reload.dylib` and `libmono-component-diagnostics_tracing.dylib`  and then the xamarin-macios embedder will need to use `mono_dl_fallback_register` to intercept dlopen and:
- for debug apps - load the components from the .framework
- for release apps - ignore the components inside the .framework and return NULL

Reply:
> but it will include all the components statically as part of it.

@lambdageek I think a single _large_ framework is _fine_, at least for the short term, as most release builds will not use frameworks anyway (to keep size down).

It would be nice to have a Debug and Release version of the framework. They could have different components enabled and also use different sets of options/optimizations better suite to debug (or not).

Having more control inside the framework sounds nice but I can't really give you a scenario right now where it would be more helpful (than debug/release). OTOH I don't know all plans and features related to the componentization of the runtime...

Reply:
> @rolfbjarne you mentioned that (limit) earlier but I can't find a link, got anything more ?

"So, I have an app here that currently has 26 dylibs, And it's taking 240 milliseconds just to load those, but if I change it and merge those dylibs into two dylibs, then it only takes 20 milliseconds to load the dylibs. So I can still have dylibs, I can still use them to share, functionality between my app and my extension, but, limiting them will be very useful. And I understand this is a tradeoff you're making between your development convenience and your application launch time for your users. Because the more dylibs that you have the easier it is to build and re-link your app in and the faster your development cycles are.

So you absolutely can and should use some, but it's good to try to target a limited number, we would, I would say off hand, a good target's about a half a dozen."

From: https://developer.apple.com/videos/play/wwdc2016/406/ (I highly recommend viewing / reading this talk, it's quite interesting).

Reply:
> There's a limit (at least a suggested one, never seen a rejection) to the number of user frameworks that an app should have. However a framework can have several dynamic libraries inside it... so we could control what's in or out.

The limit is based on performance: the more dynamic libraries (either in the form of *.dylib or *.framework) the slower startup will be. It doesn't matter where those dynamic libraries live inside the .app.

> However a framework can have several dynamic libraries inside it...

Do you mean we can put *.dylib files inside a *.framework? Or that we can embed a *.framework inside another *.framework?

Reply:
Given there are quite a number of conditions and cases I wonder if it'd make more sense for Xamarin.iOS to create the .framework as part of the build when it is needed with just the right components? E.g. I've heard that we have some requirements to not ship the diagnostics TCP server code in release builds, even if it is not used.

Reply:
We can create Mono.framework, but afaik we can't create the corresponding dSYM, that has to be done on the machine that compiled the object files.

Reply:
I've done some work relating to debug symbol generation in https://github.com/dotnet/runtime/pull/53240

dsyms for anything we generate and put in a nupkg should be available from the MS debug symbol service

Reply:
@rolfbjarne we do have the dSYM's and can put them into the runtime pack (or the symbols nuget), would you mind trying if you can create the .framework based off of that?

Reply:
@akoeplinger I created a very simple test case, and I was able to create a `.framework` and the corresponding `.framework.dSYM` from a `.dylib` and its corresponding `.dylib.dSYM`. I ran an executable that consumed the `.framework` in lldb, and lldb was able to symbolicate functions in the `.framework`. I wasn't able to make macOS' crash reporter symbolicate the framework functions though, so I'm not sure if the end result is actually correct, or if it just happens to work in lldb. One point is that Apple identifies .dSYM directories using a UUID (which shows up in the crash report), and if we create the `.framework.dSYM` from the `.dylib.dSYM`, those share the same UUID, and potentially things break.